flexible grid to streamline affine generation and pathlength function

kesshijordan opened the pull request:
Included are two related additions :

NF: The path_length function computes the shortest path, along any streamline, between the roi and each voxel. @MrBago and I are using this for a particular project we're working on, but it could be useful in other applications.

NF: The other supporting functions (flexi tvis affine, etc.) are used in path_length, but also address a problem that I've found particularly difficult as a going-up-the-learning-curve user: mapping between grid and streamline spaces with different voxel orders. After the change to using affines instead of providing voxel sizes to functions like target, one has to sort through a lot of information in the headers to get everything in the same space. These flexi tvis affine functions are extremely helpful for painless targeting and fvtk renderings.

Garyfallidis commented on dipy/tracking/utils.py:
```diff
@@ -956,3 +958,124 @@ def reduce_rois(rois, include):
+
+def flexi_tvis_affine(sl_vox_order, grid_affine, dim, voxel_size):
+    """ Computes the mapping from voxel indices to streamline points,
+        reconciling streamlines and grids with different voxel orders
```
indent to the left

kesshijordan replied:
What do you mean? I think the def is against the left edge, and the docstring is indented like the rest of the functions in this file?

Garyfallidis commented on dipy/tracking/utils.py:
```diff
@@ -956,3 +958,124 @@ def reduce_rois(rois, include):
+
+    Parameters
+    ----------
+    sl_vox_order: string of length 3
```
add space after parameter and before ``:``
please change everywhere 
it should be ```sl_vox_order : string```

kesshijordan replied:
fixed

Garyfallidis commented on dipy/tracking/utils.py:
```diff
@@ -956,3 +958,124 @@ def reduce_rois(rois, include):
+    ----------
+    sl_vox_order: string of length 3
+        a string that describes the voxel order of the streamlines (ex: LPS)
+    grid_affine: nii_aff: array (4, 4),
```
two parameters?

kesshijordan replied:
I'm interpreting this as "why do you include both the sl_vox_order and the grid_affine as parameters?" The streamline voxel order doesn't necessarily match that of the affine of the underlying grid one is trying to target/display in fvtk... the function called here (get_flexi_tvis_affine) converts both to orientations, then reconciles them with an affine.  The orientations can be different, so must be pulled from both the nifti and the trk headers.

Garyfallidis commented on dipy/tracking/utils.py:
```diff
@@ -956,3 +958,124 @@ def reduce_rois(rois, include):
+        a string that describes the voxel order of the streamlines (ex: LPS)
+    grid_affine: nii_aff: array (4, 4),
+        An affine matrix describing the current space of the grid in relation to RAS+ scanner space
+    dim: tuple of length 3
```
add space (as previously suggested)

kesshijordan replied:
fixed

Garyfallidis commented on dipy/tracking/utils.py:
```diff
@@ -956,3 +958,124 @@ def reduce_rois(rois, include):
+        An affine matrix describing the current space of the grid in relation to RAS+ scanner space
+    dim: tuple of length 3
+        dimension of the grid
+    voxel_size: array (3,0)
```
same

kesshijordan replied:
fixed

Garyfallidis commented on dipy/tracking/utils.py:
```diff
@@ -956,3 +958,124 @@ def reduce_rois(rois, include):
+
+    Returns
+    -------
+    flexi_tvis_aff: this affine maps between a grid and a trackvis space
```
same same

kesshijordan replied:
fixed fixed

Garyfallidis commented on dipy/tracking/utils.py:
```diff
@@ -956,3 +958,124 @@ def reduce_rois(rois, include):
+
+    Parameters
+    ----------
+    tvis_hdr: header from a trackvis file
```
same

Garyfallidis replied:
same... (change everywhere in your files)

kesshijordan replied:
fixed

Garyfallidis commented:
Also you may want to look at the new streamline API available in DIPY. This should solve many different problems with dealing with streamline datasets.

arokem commented:
On Sat, Sep 10, 2016 at 9:43 PM, Eleftherios Garyfallidis <
notifications@github.com> wrote:

> Also you may want to look at the new streamline API available in DIPY.
> This should solve many different problems with dealing with streamline
> datasets.
>
>
Do you mean the new streamline API in *nibabel*? Worth taking a look at
that, for sure:
http://nipy.org/nibabel/reference/nibabel.streamlines.html#module-nibabel.streamlines




> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/nipy/dipy/pull/1114#issuecomment-246161632>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAHPNlX4a1k_dZpquyOK1GRhzxf34EiPks5qo4bogaJpZM4JzTI7>
> .
>

kesshijordan commented:
Thank you for the comments. I have fixed most of the formatting issues (not sure I know what you mean regarding indentation), and commented on the number of parameters question. I also noticed that there are supposed to be two lines between each function, so I corrected that, as well. I will check out the nibabel API to see if that fixes some of these issues.

arokem commented:
Hey @kesshijordan : do you understand why these test failures are happening?

kesshijordan commented:
Hi, @arokem. Thanks for checking in. I think I figured it out: I found errors I recognized when I scrolled to the bottom of the Travis build outputs. I addressed them, resubmitted to Travis, and only one build failed (#1.5; I think because of the numpy version) "AttributeError: 'numpy.ufunc' object has no attribute 'at'", referring to a use of np.minimum.at(a,(i,j,k),b)... so I replaced it with a[i,j,k]=np.minimum(a[i,j,k],b) so that all of the builds work. Is there anything else I should be looking for besides the big ERRORs?

kesshijordan commented:
@MrBago just pointed out that substituting a[i,j,k]=np.minimum(a[i,j,k],b) for np.minimum.at(a,(i,j,k),b) introduces errors and recommended fixing it with a little function that replaces np.minimum.at if the numpy version is less than 8.

arokem commented on dipy/tracking/utils.py:
```diff
@@ -956,3 +958,149 @@ def reduce_rois(rois, include):
+    sl_vox_order : string of length 3
+        a string that describes the voxel order of the streamlines (ex: LPS)
+    grid_affine : array (4, 4),
+        An affine matrix describing the current space of the grid in relation to RAS+ scanner space
```
PEP8

arokem replied:
Line's too long, I believe.

arokem commented on dipy/tracking/utils.py:
```diff
@@ -956,3 +958,149 @@ def reduce_rois(rois, include):
+
+    sl_ornt = orientation_from_string(str(sl_vox_order))
+    grid_ornt = nib.io_orientation(grid_affine)
+    reorder_grid = reorder_voxels_affine(grid_ornt, sl_ornt, np.array(dim)-1, np.array([1,1,1]))
```
Here as well: line is longer than 80 characters (PEP8)

arokem commented on dipy/tracking/utils.py:
```diff
@@ -956,3 +958,149 @@ def reduce_rois(rois, include):
+    ----------
+    tvis_hdr : header from a trackvis file
+    nii_aff : array (4, 4),
+        An affine matrix describing the current space of the grid in relation to RAS+ scanner space
```
PEP8 (line length)

arokem commented on dipy/tracking/utils.py:
```diff
@@ -956,3 +958,149 @@ def reduce_rois(rois, include):
+    nii_aff : array (4, 4),
+        An affine matrix describing the current space of the grid in relation to RAS+ scanner space
+    nii_data : nd array
+        3D array, each with shape (x, y, z) corresponding to the shape of the brain volume,
```
Period instead of comma at the end of the sentence?

arokem commented on dipy/tracking/tests/test_utils.py:
```diff
@@ -608,3 +608,98 @@ def test_reduce_rois():
+
+    origin = np.dot(affine, [0, 0, 0, 1])
+    assert_array_almost_equal(origin[:3], np.multiply(dim,voxel_size)-voxel_size/2)
+
```
One more linebreak here (PEP8)

arokem commented on dipy/tracking/tests/test_utils.py:
```diff
@@ -608,3 +608,98 @@ def test_reduce_rois():
+    trk_point = np.dot(affine, np.append(vox_point, 1))
+
+    assert_array_almost_equal(trk_point[:3],  (vox_point[[1, 2, 0]] + 0.5) * vsz)
+
```
One more linebreak here (PEP8)

arokem commented:
Thanks! Please do consider adding documentation examples using these functions, so that we can see how exactly this is useful to you. If you need a hand with that, just let me know.

kesshijordan commented:
Sorry this took so long, @arokem. I just made a PR for the documentation that goes with this #1631.
